### PR TITLE
Interaction

### PR DESCRIPTION
**Pull request by cuongtm95**
### Script for all interactable objects in the game (like items, enemies, animals, treasure, etc..)

- [x] **Created a Interactable Base script called "Interactable", that will do the following:**

- Character will stop at a distance in front of the object (while facing it) to perform action (instead of going through/inside it)
- There's a **virtual** function inside called **Interact()** which is meant to be overwritten from other scripts to perform a specific action (For example: Open crate, Chop tree, Cook food, Make fire...)

- [x] **Pick up item and add to Inventory**

- [ ] **Show tooltip when hover over objects**

- [ ]  ...

### CODE REVIEW

**Comment by SamuelllG** [comment]
Why are @Kaes3kuch3n 's commits in here? We have a separate PR for this

**Review by SamuelllG — CHANGES_REQUESTED** [review]
Please have a look at our coding conventions here https://paper.dropbox.com/doc/General-Project-Conventions--ApoocCghuHUWZggkaloUj0IZAg-Z8kbXCDr3XqW11DkCtUHe or look at approved PRs and change a few things in your code. 

Please make Interactable an abstract class / interface. 
For the PlayerController, there is new stuff similar to yours (**as soon as someone approves the Attack_System PR**) that you can integrate the Interactable stuff into (see OnRightClick method). You'll have to pull master into your branch and battle with the merge conflicts :)
Methods for chasing enemies, IMovable etc. are already done. I guess items won't move. Depending on how much "logic" we need for interacting, you could create a class like InteractLogic, that you add to the player and call in PlayerController's OnRightClick (mentioned above). For now it's really just picking the item up, isn't it? So for now it might be enough to do that in a short method in PlayerController.

**Comment by cuongtm95** [comment]
> Please have a look at our coding conventions here https://paper.dropbox.com/doc/General-Project-Conventions--ApoocCghuHUWZggkaloUj0IZAg-Z8kbXCDr3XqW11DkCtUHe or look at approved PRs and change a few things in your code.
> 
> Please make Interactable an abstract class / interface.
> For the PlayerController, there is new stuff similar to yours (**as soon as someone approves the Attack_System PR**) that you can integrate the Interactable stuff into (see OnRightClick method). You'll have to pull master into your branch and battle with the merge conflicts :)
> Methods for chasing enemies, IMovable etc. are already done. I guess items won't move. Depending on how much "logic" we need for interacting, you could create a class like InteractLogic, that you add to the player and call in PlayerController's OnRightClick (mentioned above). For now it's really just picking the item up, isn't it? So for now it might be enough to do that in a short method in PlayerController.

Gotcha, gonna deal with it right away.

**Comment by SamuelllG** [comment]
Sorry @cuongtm95 the merge was meant to be as soon as Attack_system is merged into master.. it contains the logic and methods that I talked about in my previous comment. I'm already trying to reach @Kaes3kuch3n to approve it.
As soon as it is merged, please merge again. I will do a review afterwards.

**Comment by SamuelllG** [comment]
Merged now!

**Comment by cuongtm95** [comment]
> Merged now!

Yeah was just cleaning up the code and fix some conflicts. Will do it tonight when I get home.

**Comment by cuongtm95** [comment]
The feature is now fully working ! Waiting on review..

File: Assets/Scripts/Interactables/Crate.cs
Comment (Kaes3kuch3n) [inline Assets/Scripts/Interactables/Crate.cs]: Looks like this isn't used

File: Assets/Scripts/Interactables/Interactable.cs
Comment (Kaes3kuch3n) [inline Assets/Scripts/Interactables/Interactable.cs]: If you use `[SerializeField]`, you don't have to define the field as public

File: Assets/Scripts/Interactables/Interactable.cs
Comment (Kaes3kuch3n) [inline Assets/Scripts/Interactables/Interactable.cs]: Please add access modifiers for all fields :) (`isFocus`, `hasInteracted` and `player`)

File: Assets/Scripts/Interactables/Interactable.cs
Comment (Kaes3kuch3n) [inline Assets/Scripts/Interactables/Interactable.cs]: Please also add an access modifier here :) Should be private, I guess

File: Assets/Scripts/Interactables/Interactable.cs
Comment (Kaes3kuch3n) [inline Assets/Scripts/Interactables/Interactable.cs]: Since this method isn't implemented in here, it would make more sense to make this class and this method abstract, so that overriding it is obligatory

File: Assets/Scripts/Interactables/ItemPickup.cs
Comment (Kaes3kuch3n) [inline Assets/Scripts/Interactables/ItemPickup.cs]: You can use the player manager to get the player reference, and you can access the inventory field of the player if you create a getter for it :) And the inventory field doesn't have to be public if you get the reference yourself

File: Assets/Scripts/Player/PlayerController.cs
Comment (Kaes3kuch3n) [inline Assets/Scripts/Player/PlayerController.cs]: The nav mesh agent should stop automatically if it's close enough to the target. Changing the stopping distance manually isn't the best idea.

File: Assets/Scripts/Player/PlayerController.cs
Comment (Kaes3kuch3n) [inline Assets/Scripts/Player/PlayerController.cs]: I think this is something that can be combined with the regular movement logic, since it's basically the same

File: Assets/Scripts/Player/PlayerController.cs
Comment (Kaes3kuch3n) [inline Assets/Scripts/Player/PlayerController.cs]: We already have a FaceTarget method. I think you could also just use the existing one :)

File: Assets/Scripts/UI/TooltipDisplay.cs
Comment (Kaes3kuch3n) [inline Assets/Scripts/UI/TooltipDisplay.cs]: GetComponentInChildren is not really performant. Since the fields for this are public, it would be better if you would just reference them in the inspector

File: Assets/Scripts/UI/TooltipDisplay.cs
Comment (Kaes3kuch3n) [inline Assets/Scripts/UI/TooltipDisplay.cs]: I think it would be better to move this logic to `OnMouseOver` and `OnMouseExit`, since otherwise it is called every time in `Update()`

**Review by Kaes3kuch3n — CHANGES_REQUESTED** [review]

File: Assets/Scripts/Player/PlayerController.cs
Comment (cuongtm95) [reply Assets/Scripts/Player/PlayerController.cs]: This is a "naming" mistake from me. Actually this will make the character not targeting the object anymore and change the stopping distance to 0 so the character can move all the way to where the mouse click without stopping at a distance.

File: Assets/Scripts/Player/PlayerController.cs
Comment (cuongtm95) [reply Assets/Scripts/Player/PlayerController.cs]: Same as above, this should be called SetTarget instead of MovingToTarget. Sorry. This will make the character stop when it is close enough to the target by setting the stopping distance = the "radius" value of the object (can be tweaked)

File: Assets/Scripts/Interactables/Crate.cs
Comment (cuongtm95) [reply Assets/Scripts/Interactables/Crate.cs]: Oh yeah, already deleted !

File: Assets/Scripts/Interactables/Interactable.cs
Comment (cuongtm95) [reply Assets/Scripts/Interactables/Interactable.cs]: I have to define it as public so i can access the value in PlayerController.

File: Assets/Scripts/Interactables/Interactable.cs
Comment (cuongtm95) [reply Assets/Scripts/Interactables/Interactable.cs]: done

File: Assets/Scripts/Interactables/Interactable.cs
Comment (cuongtm95) [reply Assets/Scripts/Interactables/Interactable.cs]: I think you're right. At first i want to have something in the interact method here like "play sound, or visual or something" but maybe i will just make it abstract.

File: Assets/Scripts/Interactables/ItemPickup.cs
Comment (cuongtm95) [reply Assets/Scripts/Interactables/ItemPickup.cs]: oh ya. i forgot about PlayerManager. Already done.

File: Assets/Scripts/UI/TooltipDisplay.cs
Comment (cuongtm95) [reply Assets/Scripts/UI/TooltipDisplay.cs]: Yeah i'm consider changing the whole class and methods for Tooltip. Will update soon.

File: Assets/Scripts/Interactables/Interactable.cs
Comment (Kaes3kuch3n) [reply Assets/Scripts/Interactables/Interactable.cs]: If that's the case, you don't have to add the [SerializeField] attribute. But since I guess the player controller shouldn't be able to change the radius, it would be even better if you created a getter for the field and make it private:
```c#
public float Radius => radius;
```

File: Assets/Scripts/Interactables/Interactable.cs
Comment (Kaes3kuch3n) [reply Assets/Scripts/Interactables/Interactable.cs]: Yeah, I think it can be abstract for now. If we want to add some general implementation later on, we can still change it back 👍

**Review by Kaes3kuch3n — CHANGES_REQUESTED** [review]

File: Assets/Scripts/Player/PlayerController.cs
Comment (cuongtm95) [reply Assets/Scripts/Player/PlayerController.cs]: Okay !

File: Assets/Scripts/UI/TooltipDisplay.cs
Comment (cuongtm95) [reply Assets/Scripts/UI/TooltipDisplay.cs]: For now i will remove the tooltip script and stuffs. Will make a new PR for that.

File: Assets/Scripts/Interactables/Interactable.cs
Comment (cuongtm95) [reply Assets/Scripts/Interactables/Interactable.cs]: done

**Comment by cuongtm95** [comment]
For now i will remove the tooltip script and stuffs. Will make a new PR for that.
Player now can use Right Click to pick up items and it will be automatically added to Inventory.
Player now can use Right Click to interact with any Object that had Interactable.cs attached to it (i added Tree.cs for later purpose), for example: axe, crate, tree, enemies.. 
We can use this to implement many things in the future by overriding **Interact()** since it's a abstract method, like "play Sound" or "play animation", etc..)

File: Assets/Scripts/Interactables/Crate.cs
Comment (SamuelllG) [inline Assets/Scripts/Interactables/Crate.cs:8]: access modifiers

File: Assets/Scripts/Interactables/Crate.cs
Comment (SamuelllG) [inline Assets/Scripts/Interactables/Crate.cs]: As far as I learned from @DavidPgl it should be in Awake or maybe a SerializeField?

File: Assets/Scripts/Interactables/Interactable.cs
Comment (SamuelllG) [inline Assets/Scripts/Interactables/Interactable.cs]: please don't put values here

File: Assets/Scripts/Managers/PlayerManager.cs
Comment (SamuelllG) [inline Assets/Scripts/Managers/PlayerManager.cs]: why make it SerializeField if you use GetCompInChildren?

Shouldn't the PlayerManager only have the player anyway? I think you should get the inventory via the player that you get from the playermanager. Correct me, if I'm wrong

File: Assets/Scripts/Player/PlayerController.cs
Comment (SamuelllG) [inline Assets/Scripts/Player/PlayerController.cs]: you should also defocus when a Damageable is clicked

File: Assets/Scripts/Player/PlayerController.cs
Comment (SamuelllG) [inline Assets/Scripts/Player/PlayerController.cs]: pls put all of this in a class InteractLogic

File: Assets/Scripts/Player/PlayerController.cs
Comment (SamuelllG) [inline Assets/Scripts/Player/PlayerController.cs]: could you make the methods private?

File: Assets/Scripts/Player/PlayerController.cs
Comment (SamuelllG) [inline Assets/Scripts/Player/PlayerController.cs]: you set the stoppingDistance to 0 😮 ?

File: Assets/Scripts/Player/PlayerController.cs
Comment (SamuelllG) [inline Assets/Scripts/Player/PlayerController.cs]: that's kinda unclear. If there is no damageable, then move, after that you check for an interactable (when it's already moving). It should be if there is no damageable, check for an interactable, if there is no interactable, then move. You should include moving towards the interactable in your InteractLogic class like done for attacking.
I suggest something like:
```
Interactable interactable = objectHit.GetComponent<Interactable>();
IDamageable damageable = objectHit.GetComponent<IDamageable>();

if (damageable != null)
{
     _interactLogic.RemoveFocus();
     _attackLogic.StartAttack(objectHit);
}
else if (interactable != null)
{
     _attackLogic.StopAttack();
     _interactLogic.SetFocus(interactable);
}
else if (Physics.Raycast(clickRay, out RaycastHit groundHit, 10000, moveClickLayers))
{
     _attackLogic.StopAttack();
     _interactLogic.RemoveFocus();
     Move(groundHit);
}

```

File: Assets/Scripts/Player/PlayerController.cs
Comment (SamuelllG) [inline Assets/Scripts/Player/PlayerController.cs]: why do you set updateRotation to false below? This line is just needed because of that.

Comment (SamuelllG) [reply Assets/Scripts/Player/PlayerController.cs]: with implementation of an InteractLogic class not needed anymore

File: Assets/Scripts/Interactables/ItemPickup.cs
Comment (SamuelllG) [inline Assets/Scripts/Interactables/ItemPickup.cs]: Can an item be picked up multiple times, for example after throwing it away from inventory? Then I think you should get the inventory once, and then always add to it. Don't get it via the playermanager > player every time the item is picked up.

**Review by SamuelllG — CHANGES_REQUESTED** [review]
I suggest implementing an InteractLogic class :)

File: Assets/Scripts/Interactables/Crate.cs
Comment (cuongtm95) [reply Assets/Scripts/Interactables/Crate.cs:8]: is done !

File: Assets/Scripts/Interactables/Crate.cs
Comment (cuongtm95) [reply Assets/Scripts/Interactables/Crate.cs]: Alright ! changed it to Awake()

File: Assets/Scripts/Interactables/Interactable.cs
Comment (cuongtm95) [reply Assets/Scripts/Interactables/Interactable.cs]: I just want to set a default value for it. Will do it in Start() then.

File: Assets/Scripts/Managers/PlayerManager.cs
Comment (cuongtm95) [reply Assets/Scripts/Managers/PlayerManager.cs]: I'm not sure if PlayerManager should only have the player. If so i will create a getter for inventory in other place.

File: Assets/Scripts/Player/PlayerController.cs
Comment (cuongtm95) [reply Assets/Scripts/Player/PlayerController.cs]: Oki !

File: Assets/Scripts/Player/PlayerController.cs
Comment (cuongtm95) [reply Assets/Scripts/Player/PlayerController.cs]: is done

File: Assets/Scripts/Player/PlayerController.cs
Comment (cuongtm95) [reply Assets/Scripts/Player/PlayerController.cs]: this looks good to me.

File: Assets/Scripts/Player/PlayerController.cs
Comment (cuongtm95) [reply Assets/Scripts/Player/PlayerController.cs]: This basically mean that if the character not targeting anything then it will be switched back to normal, that means the character will not stop before reaching the position where the mouse clicked.

File: Assets/Scripts/Player/PlayerController.cs
Comment (cuongtm95) [reply Assets/Scripts/Player/PlayerController.cs]: Yeah just removed updateRotation. Seems like it is not needed

File: Assets/Scripts/Interactables/ItemPickup.cs
Comment (cuongtm95) [reply Assets/Scripts/Interactables/ItemPickup.cs]: Yeah changed to this 
```suggestion
        private void Awake()
    {
        GetInventory();
    }

    private Inventory GetInventory()
    {
        return inventory = PlayerManager.Instance.GetPlayer().GetComponentInChildren<Inventory>();
    }
```
should be working fine

Comment (cuongtm95) [reply Assets/Scripts/Interactables/ItemPickup.cs]: For now since we are not able to drop the item yet so i can't test it but im pretty sure it will work just fine.

**Comment by cuongtm95** [comment]
- Moved everything Interact-related from PlayerController to InteractLogic.cs and cleaned some unneeded code.
- Get Inventory once on ItemPickup at Awake()

File: Assets/Scripts/InteractLogic.cs
Comment (SamuelllG) [inline Assets/Scripts/InteractLogic.cs]: We don't have IMovable anymore, if you pull master to resolve merge conflicts, you'll see. Please resolve merge conflicts

File: Assets/Scripts/Interactables/Interactable.cs
Comment (SamuelllG) [reply Assets/Scripts/Interactables/Interactable.cs]: thanks!

File: Assets/Scripts/Managers/PlayerManager.cs
Comment (SamuelllG) [reply Assets/Scripts/Managers/PlayerManager.cs]: @tobihoffmann @DavidPgl @Kaes3kuch3n @potatoTVnet @thuemmels

File: Assets/Scripts/Player/PlayerController.cs
Comment (SamuelllG) [reply Assets/Scripts/Player/PlayerController.cs]: You're right, default stoppingDistance is 0. I confused things here!

File: Assets/Scripts/Interactables/ItemPickup.cs
Comment (SamuelllG) [reply Assets/Scripts/Interactables/ItemPickup.cs]: thanks!

File: Assets/Scripts/InteractLogic.cs
Comment (SamuelllG) [inline Assets/Scripts/InteractLogic.cs]: You could think about giving Movable a public method SetStoppingDistance. In my opinion it's better to do it all via the Movable instead of getting the NavMeshAgent here.

File: Assets/Scripts/InteractLogic.cs
Comment (SamuelllG) [inline Assets/Scripts/InteractLogic.cs]: Is there a reason for * .8f ?

File: Assets/Scripts/InteractLogic.cs
Comment (SamuelllG) [inline Assets/Scripts/InteractLogic.cs]: Should focus and target be SerializeFields? Are they meant to be set via inspector?

File: Assets/Scripts/Player/PlayerController.cs
Comment (SamuelllG) [inline Assets/Scripts/Player/PlayerController.cs]: Don't forget to RemoveFocus here :) or was it on purpose?

File: Assets/Scripts/Player/PlayerController.cs
Comment (SamuelllG) [inline Assets/Scripts/Player/PlayerController.cs]: objectHit.GetComp
It's just a minor thing, but it makes it a little bit faster I guess :]

File: Assets/Scripts/Player/PlayerController.cs
Comment (SamuelllG) [reply Assets/Scripts/Player/PlayerController.cs]: Thanks for implementing! with interactlogic it looks much better 👍

**Review by SamuelllG — CHANGES_REQUESTED** [review]
I commented a few things. Please resolve merge conflicts :)

File: Assets/Scripts/InteractLogic.cs
Comment (cuongtm95) [reply Assets/Scripts/InteractLogic.cs]: Just for us to see via inspector what the object is. Nothing need to be set here.

File: Assets/Scripts/InteractLogic.cs
Comment (cuongtm95) [reply Assets/Scripts/InteractLogic.cs]: No actually. I just hard coded it and forgot to change back.

File: Assets/Scripts/Player/PlayerController.cs
Comment (cuongtm95) [reply Assets/Scripts/Player/PlayerController.cs]: Oh yeah i want to also focus on the damageable as well, so addition to attack we can do something more if we attach Interactable.cs on it. Without Interactable.cs nothing will be focused.

File: Assets/Scripts/Player/PlayerController.cs
Comment (cuongtm95) [reply Assets/Scripts/Player/PlayerController.cs]: gotcha

File: Assets/Scripts/InteractLogic.cs
Comment (cuongtm95) [reply Assets/Scripts/InteractLogic.cs]: is done !
```suggestion
        public void SetStoppingDistance(float distance)
        {
            NavMeshAgent.stoppingDistance = distance;
        }
```

Comment (cuongtm95) [reply Assets/Scripts/InteractLogic.cs]:
```suggestion
        public void SetStoppingDistance(float distance)
        {
            NavMeshAgent.stoppingDistance = distance;
        }
```

Comment (cuongtm95) [reply Assets/Scripts/InteractLogic.cs]: is done !
```suggestion
        public void SetStoppingDistance(float distance)
        {
            NavMeshAgent.stoppingDistance = distance;
        }
```

```suggestion
       _movable.SetStoppingDistance(newTarget.Radius);
```

File: Assets/Scripts/Entity/Player/PlayerController.cs
Comment (Kaes3kuch3n) [inline Assets/Scripts/Entity/Player/PlayerController.cs]:
```suggestion
                else if (objectHit.GetComponent<Interactable>() is Interactable interactable)
```
I'd rewrite this a little so that the variable declaration and the null check are joined together, because we don't need the variables outside of this scope, so this would be a little clearer :) And if you're already on it, can you change the damageable check the same way? 😅

File: Assets/Scripts/InteractLogic.cs
Comment (Kaes3kuch3n) [reply Assets/Scripts/InteractLogic.cs]: If you just need to see them in the inspector, you can also remove the SerializeField attributes and simply switch the inspector to debug mode. In this mode it also displays the private fields that don't have the attribute

File: Assets/Scripts/Interactables/ItemPickup.cs
Comment (Kaes3kuch3n) [inline Assets/Scripts/Interactables/ItemPickup.cs]: I think this isn't really needed, since you can simply drag in the reference to the inventory into the serialised field :)

**Review by Kaes3kuch3n — CHANGES_REQUESTED** [review]
Almost there, just a few small change requests left :)

File: Assets/Scripts/Interactables/ItemPickup.cs
Comment (cuongtm95) [reply Assets/Scripts/Interactables/ItemPickup.cs]: Oh okay. I thought we don't want to manually assign anything so...

File: Assets/Scripts/Entity/Player/PlayerController.cs
Comment (cuongtm95) [reply Assets/Scripts/Entity/Player/PlayerController.cs]: Alright !

**Comment by cuongtm95** [comment]
Should be okay now.

**Comment by cuongtm95** [comment]
Should be fine now. Please check again and merge to master without GameScene

**Review by Kaes3kuch3n — DISMISSED** [review]

File: Assets/Scripts/Player/PlayerController.cs
Comment (SamuelllG) [reply Assets/Scripts/Player/PlayerController.cs]: I think you misunderstood me, @Kaes3kuch3n updated the code

**Review by SamuelllG — DISMISSED** [review]

**Review by inkompotato — APPROVED** [review]
tested by Luis - trusted by me

**Review by SamuelllG — APPROVED** [review]

**Pull request merged**
